helix-term/src/ui/editor.rs

```diff
@@ -697,6 +746,370 @@ impl EditorView {
+                        additional_area,
+                    );
+                    new_offset.anchor = node.byte_range.end;
+                    let highlights = Self::doc_syntax_highlights(doc, new_offset.anchor, 1, theme);
+                    render_text(
+                        &mut renderer,
+                        text,
+                        new_offset,
+                        &doc.text_format(additional_area.width, Some(theme)),
+                        doc_annotations,
+                        highlights,
+                        theme,
+                        line_decoration,
+                        translated_positions,
+                    );
+
+                    // draw the "..." with the keyword.operator style
+                    surface.set_stringn(
+                        (already_written + line_context_area.x).saturating_sub(1),
+                        additional_area.y,
+                        dots,
+                        dots.len(),
+                        theme.get("keyword.operator"),
+                    );
+                }
+
+                let line_num_anchor = node.byte_range.start;
+
+                // get all highlights from the latest point
+                let highlights = Self::doc_syntax_highlights(doc, line_num_anchor, 1, theme);
+
+                let mut renderer = TextRenderer::new(
+                    surface,
+                    doc,
+                    theme,
+                    view.offset.horizontal_offset,
+                    line_context_area,
+                );
+                new_offset.anchor = line_num_anchor;
+
+                // limit the width to its size - 1, so that it won't draw trailing whitespace characters
+                line_context_area.width = already_written - 1;
+
+                render_text(
+                    &mut renderer,
+                    text,
+                    new_offset,
+                    &doc.text_format(line_context_area.width, Some(theme)),
+                    doc_annotations,
+                    highlights,
+                    theme,
+                    line_decoration,
+                    translated_positions,
+                );
+
+                // next node
+                context_area.y += 1;
+            }
+        }
+    }
+
+    fn get_node_range<'a>(
+        query_nodes: QueryMatches<'a, 'a, RopeProvider<'a>>,
+        start_index: u32,
+        end_index: u32,
+        top_first_byte: usize,
+        cursor_byte: usize,
+    ) -> Option<std::ops::Range<usize>> {
+        query_nodes
+            .flat_map(move |qnode| {
+                let end_nodes: Vec<_> = qnode.nodes_for_capture_index(end_index).collect();
+                qnode
+                    .nodes_for_capture_index(start_index)
+                    .flat_map(move |context| {
+                        end_nodes
+                            .clone()
+                            .into_iter()
+                            .filter(move |it| {
```

Comment:
```suggestion
                    .flat_map(|context| {
                        end_nodes
                            .iter()
                            .filter(move |it| {
```
We don't need the clone here.

Reply:
it's necessary, because otherwise this becomes a `FnMut` closure.

Also; the pullrequest is far from ready yet, and i still haven't refactored the code, so please bear with me :)